[Low Joe Yi] iP

JJoeYi opened the pull request:
# DUKE header

- point 1
- point 2

1. point1
2. point2

`code block`

- [ ] task list
- [x] 2 done
- [ ] 3

🈂️ 🈂️ 

> quote block

[hyperlink](https://github.com/nus-cs2103-AY2223S1/ip/pull/280)

__bold__ _italic_ ~~strikethrough~~

JJoeYi committed:
Level-1, edited Duke.java

JJoeYi committed:
Level-2, edited Duke.java

JJoeYi committed:
Level-3, added Task class

JJoeYi committed:
Level-4, added Todo, Deadline, Event java files and fixed list command's formatting

JJoeYi committed:
A-TextUiTesting

JJoeYi committed:
Level-5. Removed past level codes + added DukeException

JJoeYi committed:
Level-6, added deleting capability

JJoeYi committed:
Add saving functionality to Duke, Duke can now read / write save file

JJoeYi committed:
add LocalDate to Deadline class, level-8

JJoeYi committed:
Merge branch 'branch-Level-7'

JJoeYi committed:
Merge branch 'branch-Level-8'

JJoeYi committed:
add more classes to Duke, allows more encapsulation. A-MoreOOP

JJoeYi committed:
move files to "duke" package

JJoeYi committed:
Add Gradle to project, A-Gradle

JJoeYi committed:
Add JUnit tests

JJoeYi committed:
change mainClassName for build.gradle for building Jar file

JJoeYi committed:
Add javadocs to all classes and methods

JJoeYi committed:
Give users a way to find a task by searching for a keyword.

JJoeYi committed:
Merge branch 'branch-A-JavaDoc'

JJoeYi committed:
Merge branch 'branch-Level-9'

JJoeYi committed:
Add GUI to Duke

JJoeYi committed:
add changes to parser to return a string instead, to display to GUI

LimAiLin commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+public class Deadline extends Task {
```
I think this is a nice application of OOP :)

GarethOng commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,66 @@
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
+        this.byDate = LocalDate.parse(by, DateTimeFormatter.ofPattern("dd-MM-yyyy"));
```
DateTimeFormatter.ofPattern("dd-MM-yyyy") can be declared as a static variable. I think it can reusable in the scenario that more formatting is needed to be done?

GarethOng commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,66 @@
+        boolean isDone = isDoneString == X;
+        int idxOfBy = s.indexOf("(by:");
+
+        String description = s.substring(7, idxOfBy);
```
Can 7 be declared as a constant? I think it will greatly improve the readability of your code!

GarethOng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @throws DukeException When command is unclear or string is given in a wrong format
+     */
+    public void parse() throws DukeException {
```
Can this be replaced with switch statements? Switch statements can reduce the number of times doing repeated checks such as (line.length <= 7) which will boost the readability of the program.

GarethOng commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,47 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public void markAsDone() {
+        this.isDone = true;
+    }
+
+    public void markAsUndone() {
+        this.isDone = false;
+    }
```
I think this can be combined into one function
```suggestion
    public void toggleDone() {
        this.isDone = !this.isDone;
    }
```

JJoeYi committed:
Merge branch 'branch-Level-10'

JJoeYi committed:
add assertions in Deadline and Event classes

assertions for Deadline and Event are added to ensure that the
strings used to create a Deadline or Event have the necessary
keywords "(by:" and "(at:" respectively. Otherwise the methods will
not work at all.

Let's add assertions to check for the existence of the specific
keywords

JJoeYi committed:
Add Command interface along with all the command classes

parse() method was too long and confusing to understand.

Make the different cases into command classes, implement from
a Command interface

The code follows SLAP principle and is of better code quality.

JJoeYi committed:
Merge pull request #1 from JJoeYi/branch-A-CodeQuality

Add Command interface along with all the command classes

JJoeYi committed:
Merge pull request #2 from JJoeYi/branch-A-Assertions

add assertions in Deadline and Event classes

JJoeYi committed:
Improve GUI and add personality to Du-kopi

JJoeYi committed:
Update the User Guide

JJoeYi committed:
Update the mainClassName for build.gradle to be launcher instead
of Duke class

JJoeYi committed:
Update README.md
